src/js/common/utils/accessible-menus.js

```diff
@@ -0,0 +1,349 @@
+    // The focusElement should be a <li>, so we want to focus on the contents
+    focusElement.firstElementChild.focus();
+  }
+}
+
+
+/**
+ * Gets the menu and associated menu button from an element if possible. Returns null if not.
+ *
+ * @param {HTMLElement} menuLi  The element (probably <li>) that may contain a menu
+ * @return {Object|null}        If the element contains a proper menu structure, returns an object
+ *                               { menu: <HTMLElement>, menuButton: <HTMLElement> }
+ *                               Otherwise, returns null
+ */
+function getMenuStructure(menuLi) {
+  const menuButton = menuLi.querySelector('button, [role="button"]');
+  const menu = menuButton ? menuLi.querySelector(`#${menuButton.getAttribute('aria-controls')}`) : null;
+
+  if (!menuButton || !menu) {
+    return null;
+  }
+  const menuRole = (menu.getAttribute('role') || '').toLowerCase();
+  if (!(menuRole === 'menu' || menu.querySelector('[role="menu"]'))) {
+    return null;
+  }
+
+  return { menuButton, menu };
+}
+
+
+/**
+ * Closes a menu.
+ *
+ * @param {HTMLLIElement} menuLI  The <li> containing the menubutton and menu
+ */
+function closeMenu(menuLi) {
+  // If we're not dealing with a menu structure, abort
+  const struct = getMenuStructure(menuLi);
+  if (!struct) {
+    return;
+  }
+
+  const { menuButton, menu } = struct;
+
+  // Close the menu
+  menuButton.removeAttribute('aria-expanded');
+  menu.setAttribute('hidden', 'hidden');
+}
+
+
+/**
+ * Opens a menu.
+ *
+ * @param {HTMLLIElement} menuLi  The <li> containing the menubutton and menu
+ * @param {bool} openSubMenu      Whether or not the first submenu should be opened. E.g. if we're opening the
+ *                                 second-level menu in a wide screen, we want the third-level opened as well
+ * @param {bool} stealFocus       Whether or not to focus on the first item in the opened menu
+ */
+function openMenu(menuLi, openSubMenu = false, stealFocus = true) {
+  // If we're not dealing with a menu structure, abort
+  const struct = getMenuStructure(menuLi);
+  if (!struct) {
+    return;
+  }
+
+  const { menuButton, menu } = struct;
+
+  // First, close all sibling menus
+  const openMenus = menuLi.parentElement.querySelectorAll('[aria-expanded=true]');
+  openMenus.forEach(m => closeMenu(m.parentElement));
+
+  // Open the menu
+  menuButton.setAttribute('aria-expanded', true);
+  menu.removeAttribute('hidden');
+
+  if (stealFocus) {
+    const menuRole = (menu.getAttribute('role') || '').toLowerCase();
+    const firstMenuItem = menuRole === 'menu' ?
+      firstVisibleChild(menu) :
+      firstVisibleChild(menu.querySelector('[role="menu"]'));
+    firstMenuItem.firstElementChild.focus();
+  }
+
+  // If we're wide-screen (for example), open the first submenu
+  if (openSubMenu) {
+    openMenu(menu.firstElementChild, false, false);
+  }
+}
+
+
+/**
+ * Creates a closure which closes all menus in the provided element.
+ *
+ * @param {HTMLElement}
```

Comment:
I assume this `{HTMLElement}` is called `menuElement`

Reply:
😛